Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/CoilUtils.kt

```diff
@@ -2,59 +2,94 @@
 
 package org.dolphinemu.dolphinemu.utils
 
+import android.graphics.drawable.Drawable
 import android.net.Uri
 import android.view.View
 import android.widget.ImageView
-import coil.load
-import coil.target.ImageViewTarget
+import coil.ImageLoader
+import coil.decode.DataSource
+import coil.executeBlocking
+import coil.fetch.DrawableResult
+import coil.fetch.FetchResult
+import coil.fetch.Fetcher
+import coil.imageLoader
+import coil.key.Keyer
+import coil.memory.MemoryCache
+import coil.request.ImageRequest
+import coil.request.Options
+import org.dolphinemu.dolphinemu.DolphinApplication
 import org.dolphinemu.dolphinemu.R
 import org.dolphinemu.dolphinemu.adapters.GameAdapter.GameViewHolder
 import org.dolphinemu.dolphinemu.features.settings.model.BooleanSetting
 import org.dolphinemu.dolphinemu.model.GameFile
 import java.io.File
 import java.io.FileNotFoundException
 
+class GameIconFetcher(
+    private val game: GameFile,
+    private val options: Options
+) : Fetcher {
+    override suspend fun fetch(): FetchResult {
+        val customCoverUri = CoilUtils.findCustomCover(game)
+        val builder = ImageRequest.Builder(DolphinApplication.getAppContext())
+        val drawable: Drawable? = if (customCoverUri != null) {
+            val request = builder.data(customCoverUri).error(R.drawable.no_banner).build()
+            DolphinApplication.getAppContext().imageLoader.executeBlocking(request).drawable
+        } else if (BooleanSetting.MAIN_USE_GAME_COVERS.boolean) {
+            val request = builder.data(
+                CoverHelper.buildGameTDBUrl(game, CoverHelper.getRegion(game))
+            ).error(R.drawable.no_banner).build()
+            DolphinApplication.getAppContext().imageLoader.executeBlocking(request).drawable
+        } else {
+            null
+        }
+
+        return DrawableResult(
+            drawable = drawable!!,
+            isSampled = false,
+            dataSource = DataSource.DISK
+        )
+    }
+
+    class Factory : Fetcher.Factory<GameFile> {
+        override fun create(data: GameFile, options: Options, imageLoader: ImageLoader): Fetcher =
+            GameIconFetcher(data, options)
+    }
+}
+
+class GameIconKeyer : Keyer<GameFile> {
+    override fun key(data: GameFile, options: Options): String = data.getPath()
```

Comment:
Might be nice to prepend the game ID (or something along those lines) to better handle the case where the user renames a game into a name previously used by a different game.

Reply:
Sure